### DISCUSSION TITLE
search tabs keyboard shorcut

### DISCUSSION THREAD

**Discussion by mr-cheffy** [zen-browser/desktop#807] [Ideas]
### What feature would you like to see?

I would love it if the browser had something like search tabs shortcut on mac cmd + shift + a. Microsoft edge has this and it's a game changer. please take a look at the edge implementation rather than the chrome implementation. The good thing about edge is that not only the search tabs bar in centered it also look through your recent history and tabs which you recently closed also show up in the search. I think this feature will be a game changer.  

`Edge Search Tabs` 

<img width="466" alt="image" src="https://github.com/user-attachments/assets/cccb0fdd-927f-4bc8-bd35-3ea4d379ea12">

### What platform are you seeing the problem on?

macOS

 
 ~ _Created automatically from issue: https://github.com/zen-browser/desktop/issues/526_

**Comment by rinormaloku** [comment]
This is the only thing that annoys me about zen (firefox) browser. 

Note: You can use QuicKey which is a nice extension and is activated with Alt + Q. In my case I swap alt and ctrl, so the shortcut is unusable AND firefox doesn't allow for customizing browser shortcuts.

**Reply by pencilcheck** [reply]
Thanks, this seems to be a good workaround, I can just do C-q and it will open up a quick search menu.

**Reply by yz47** [reply]
@rinormaloku QuicKey is also my workaround for now. However, there's a major bug, when switching tabs across different workspaces, QuicKey will hang. It's not just QuicKey, I tried a few other tab search/switch addons and they all share the same buggy behavior. So it's likely something with Zenbrowser.

**Comment by giuli635** [comment]
You can do Ctrl+L and type % to search tabs

**Reply by rinormaloku** [reply]
That's a shortcut that makes me use the mouse

**Reply by giuli635** [reply]
I don't know where is the point where you need to use the mouse, sorry, could you clarify?

**Reply by rinormaloku** [reply]
QQ: _Can you perform a search tab and share here how many hand movements and keystrokes you need to do to go to the tab you intended_? It's 7 different keys... at that point, you are more efficient using a mouse to go to the tab. 

Keep in mind we use shortcuts for productivity and efficiency, not to be cool or hip.

**Reply by giuli635** [reply]
I see, well, with Ctrl+L you are focused at the url bar, you just need to type % and a space and the search keywords you want, then (i don't like it, i would prefer other keys) you can choose from the options with the arrows.

**Reply by AH-Merii** [reply]
I use Alt + D and then type %. It's not too bad, for Edge I would have to use Ctrl + Shift + A.

**Reply by MigueloAtla** [reply]
Maybe a good solution is to have a shortcut for the Ctrl+L and % combination, similar to how Ctrl+K opens the URL bar with the search engine configured.

**Reply by apostolisanastasiou** [reply]
It doesnt work for me. My setup:
- 2 spaces, 1 work 1 personal.
- I'm searching for a tab on the same space I'm active on

https://github.com/user-attachments/assets/ceb5c089-d5f6-4788-8661-6361a774c054


- Some tabs appear on search, some don't.
- Some tabs, like "github.com" appear twice for some reason
  -  One is a dead link, does nothing when you click/enter them. I assume that's because it's the tab that's already active and I'm writing now. But it's obv a UX issue and UI should make obvious what's happening.
  - Other opens a .... new tab? When I was searching for a tab. I dont get it tbh
- Tabs show their URLs, not their names. It's hard to search like that.

**Comment by joebonneau** [comment]
I would also love this feature. It's probably my favorite feature of a browser like Arc, and it feels like a step down to my workflow to not have a direct (and customizable) hotkey to switch tabs.

**Comment by yz47** [comment]
Please please please consider implementing this!!! 

Option 1: 
- Implement a native tab search functionality

Option 2:
- Bring back "List tabs" from Firefox and allow setting a shortcut for it

Option 3:
- Allow setting a short cut to `cmd + L -> % -> space`. This should be entirely possible as it's the same as how the "Focus search" action puts  `@google -> space` in the address bar. 

We want to quickly search open tabs and navigate to it. `cmd + L -> % -> space` is three action inputs instead of just one, and `%` is not the easiest key to get to. it's essentially `cmd + L -> shift + 5 -> space`, which in practice is significantly more painful than a simple shortcut, like `cmt + shift + a` in chrome. And the fatigue adds up very quickly.

**Reply by eeriemyxi** [reply]
Something worth noting about Option 3 is that the shortcut is actually `CTRL + l + shift + 5 + space + <input> + tab`. That's because it doesn't focus on the first result by default, you will have to explicitly press tab. In other implementations, you don't have to press tab unless you want to move to the second best result.

**Comment by nistrup** [comment]
Why was the issue closed for this? https://github.com/zen-browser/desktop/issues/526

As a browser that prices itself on easy navigation with shortcuts it's very unfortunate that there is no way to easily switch between tabs using search, it effectively means that the easiest workflow is to just create a new tab with `cmd + L` or `cmd + T` even if there is already a tab open with the page you want to navigate to. It creates bloat and you end up with 50+ tabs instead of a lower number with cleaner switching / searching between them.

**Reply by rui-monte** [reply]
The issue was closed to be discussed here.

Also  I think with the new bar `cmd + T` "search with Google or enter address" could be a nice opportunity to implement this feature, this feature was the one I missed switching from Arc. I almost never used the sidebar with this feature, is super effectively for workflow

**Reply by JSGRD22** [reply]
No, the easiest workflow is to use the standard % shortcut to search tabs. The only room for improvement here is a more convenient shortcut to do this.

**Reply by yijiaz-mc** [reply]
@JSGRD22 Is there plan for "a more convenient shortcut" for this?

**Reply by rui-monte** [reply]
I understand, just a point to add, even with `%` is not the same efficiency as Arc.

The pre selected option is search with Google
![image](https://github.com/user-attachments/assets/f31822f7-aa37-424f-9dc1-98b23a8213ec)

Where the Arc the same bar pre select opened tabs
![image](https://github.com/user-attachments/assets/6cf2d1a0-e55d-41b5-aada-341f55008344)

**Reply by JSGRD22** [reply]
You need to add a space after the modifer

**Reply by geraldzm** [reply]
@JSGRD22 that is not the only room for improvement. 

Adding the shortcut and sorting the results by relevance score. Sometimes I have to type almost the entire tab title for it to appear as the first option, the standard % is just not useful.

**Comment by hinaloe** [comment]
I'm using [Quick Tabs Port for firefox](https://github.com/babyman/quick-tabs-chrome-extension/issues/189#issuecomment-2563974135). This extension provides awesome incremental search specific to tabs, and I have been using it with a Chromium-based browser for a long time.

![image](https://github.com/user-attachments/assets/672afe98-3a6f-4983-9b9a-265252807e4c)![image](https://github.com/user-attachments/assets/88b1e3bb-c0f5-4bf3-8d4c-82dda19ebcbb)


This port allows it to be used with Zen, and of course, key bindings can be changed (some key bindings did not work well. I also often use Ctrl/cmd+E for this, but it is conflicted and not available). But unfortunately it is not available in the official repositories.

Relying on extensions is one thing, but I think it should be a feature that can be natively operated with consistent usability and UI. (I'm eagerly awaiting it).

**Comment by prasoon2211** [comment]
I started using [TabSearch for Firefox](https://addons.mozilla.org/en-US/firefox/addon/tab_search/) and it works really well. Just need to remap the short (go to Extension, click on the three dots and remap the extension shortcut). I've mapped it to Opt + Cmd + Space (Cmd + Shift + A cannot be remapped because it's taken by Zen)

**Comment by ChristinWhite** [comment]
[Surfingkeys](https://addons.mozilla.org/en-US/firefox/addon/surfingkeys_ff/) also has good tab searching, it's what I'm using now, but it does a lot more for Vim enjoyers, has a learning curve though.

**Comment by kaeside** [comment]
One of the subtle yet highly useful features of Chrome’s tab search functionality is how it lets you quickly switch between your most recently visited tabs.

### Example Use Case:

The user has the following tabs open:
- Tab 1  
- Tab 2  
- Tab 3  

The user was previously on **Tab 1** and is now actively on **Tab 3**.

1. The user presses `Cmd+Shift+A` (Mac) or `Ctrl+Shift+A` (Windows/Linux).  
2. The **tab search menu opens**, and the **first result** shown is the **previously visited tab (Tab 1)**.  
3. That tab is **auto-highlighted**, so the user can simply press **Enter** to return to Tab 1.

Later, the user wants to return to Tab 3:

1. They press the same shortcut again (`Cmd+Shift+A` / `Ctrl+Shift+A`).  
2. The **tab search menu opens**, and **Tab 3** (the previously visited tab) is now listed first and auto-highlighted.  
3. Pressing **Enter** takes the user back to Tab 3.

It's a fast and seamless way to jump between two tabs without losing focus or context.

---

In comparison, Firefox’s implementation—triggered by typing `%` and then a space after `Ctrl+L`—has a few drawbacks:

1. It only shows a **limited number of tabs** (currently six), even when more are open.  
2. The list is not sorted in a useful way—neither alphabetically nor by last visited.  
   - For example, after switching to another GitHub page, I expected that to show first, but Firefox lists tabs like `about:preferences`, Reddit, Amazon, Investopedia, and StockCharts instead—regardless of my recent activity.  
3. The process requires multiple steps: `Ctrl+L`, then `%`, then space. That’s slower and more cumbersome than using a dedicated keyboard shortcut.

---

I’d love to see Zen Browser adopt a Chrome-like approach here:
- A single, dedicated shortcut to bring up tab search  
- Auto-highlight the **last visited tab**  
- Enable quick toggling between recent tabs with just **one shortcut and one keystroke** (`Enter`)  

It’s a small feature that makes a **huge difference** in productivity and user experience.

**Reply by eeriemyxi** [reply]
Firefox has an even cooler feature that you can enable by enabling Preferences > Tab > CTRL + Tab option.

That feature lets you quickly switch in recently used order. There is no enter involved. It is quicker than your Chromium solution.

This is of course not a replacement for searching tabs, nor am I suggesting that.

**Reply by vraravam** [reply]
yes - this is the reason I have never switched to Chrome or its derivatives from the beginning.

**Reply by kaeside** [reply]
Thanks that's a great workaround for some of the functionality. It works better than Chrome (I use brave) as it jumps back and forth between the tabs instead of keep cycling backwards. 

The one thing it is missing is the search so you can filter by just lets say all open Jira tabs and then close all of the Jira tabs.

**Comment by rstanuwijaya** [comment]
I'm trying to bring this functionality by this PR #8272, but it is still incomplete. 

Edit: seems like quite complete now

**Reply by rstanuwijaya** [reply]
Hi all, this is the latest demo of the search tab/tab switcher in action. I bind Alt+q to trigger the tab search mode, and added some patches to sort the tabs suggestion by recently opened order.

Feel free to add any feedback :)

Quick tab switching

https://github.com/user-attachments/assets/a010d9bc-9ca1-4f70-be37-c5c619d2376c

Tab search

https://github.com/user-attachments/assets/009c4f64-80c3-4698-99a1-08df1ce9baa2

Tab switching

https://github.com/user-attachments/assets/8c3916f6-b9f8-4c74-b542-57363ae5c6ed

**Reply by eeriemyxi** [reply]
Is there some way to apply this patch to compiled builds or something like that? It seems like the Zen Browser's official instructions need you to compile the browser engine yourself as part of the build process. I have a potato PC so I don't think that's an option for me.

**Reply by fahidsarker** [reply]
Thanks for the awesome PR. I am really frustrated with current approach. But seems like it will take a while before we get our hands on it. Is there anyway, we can apply it right now? Maybe with a Zen Mod?

**Reply by rstanuwijaya** [reply]
It's not possible to apply with zen mods, because it modifies some of the Firefox's code. Let's wait patiently for the review process :)

**Reply by jordy758** [reply]
@rstanuwijaya awesome work! I was wondering if this somehow could also be implemented on the command + t popup. I LOVED the feature in Arc where I could switch between most recent tab with just command + t + enter. At the moment there is no default option that's selected whenever you command + t, so it always requires input. But I think it's pretty cool if the default would be the most recent. WDYT?

**Reply by geraldzm** [reply]
awesome, is there some way to limit the search to the current workspace ?

**Comment by accmeboot** [comment]
It would be really great, if just could add sort order to the suggestions, so folks who need to see tabs first, could achieve that without using different shortcuts/extra typing

**Comment by camdarley** [comment]
Coming from Arc, what I miss most isn't a dedicated tab search shortcut - it's that **I never had to wonder if something was already open**. 

Having to type `Ctrl+L → % → space` or remember `Alt+Q` still forces me to think "wait, do I need to search tabs or just open a new one?"

**What if open tabs just appeared in the regular search popup?**

No special shortcuts, no `%` prefix, no decisions. Just search for what you need and Zen shows if it's already open.

This feels very "Zen" to me - remove the friction, let the browser handle the complexity.

**Reply by Seismix** [reply]
This is already a feature:

<img width="1715" height="657" alt="image" src="https://github.com/user-attachments/assets/eb2c800b-b5a1-445f-b2e6-8abb91314b76" />

Although I would prefer the default to show already opened tabs first.

Or do you mean something else?

**Reply by jordy758** [reply]
I think what would be awesome is if you press `command + t` that the default option would be the most recent tab. At this moment there is no default action, where as in Arc you could just do `command + t + enter/return` to switch between last active tab.

**Reply by Seismix** [reply]
Not on ctrl + t, but the action exists, it's ctrl + tab (with the option to cycle through recently used tabs set).

Unfortunately there is no way to rebind that as far as I know 😞

**Reply by papertree0** [reply]
For me this is what makes the most sense. Having `ctrl+t` prioritize already open tabs for easy switching.

**Reply by camdarley** [reply]
@Seismix, @papertree0  Thanks for your responses!
Indeed, I hadn't noticed the feature was already there. But still, I agree: the fact that the open tab suggestion is not at the top of the list defeats the purpose.
This would align with the seamless experience described in the original request - users wouldn't need to think about whether something is open, it would simply appear first when relevant.

**Comment by lorenoamori** [comment]
I want this feature, having to manually type `%` its not that bad, but it would be nice if the `%` could just be there.

**Comment by moiSentineL** [comment]
I made a `bash` script that I use with [sxhkd](https://github.com/baskerville/sxhkd) (for bspwm) to simulate the key combination:

```bash
#!/usr/bin/env bash

# This script checks for active window class
# and listens for commands via xdotool

# Get the active window's WM_CLASS
active_class=$(xprop -id "$(xdotool getactivewindow)" WM_CLASS)

# Check if it's Zen Browser
if [[ "$active_class" == *"zen"* ]]; then
    # Send Ctrl+L (focus address bar)

    xdotool key --clearmodifiers "ctrl+l"
    xdotool type --clearmodifiers "% "
    xdotool keyup space

    sleep 0.01

    xdotool keyup "alt" # because I have binded this script to alt + @space in sxhkd
fi
```

There are some problems though, sometimes the `alt` key stays pressed (despite the `keyup`). I guess I need to tweak it a bit more. I hope this helps!

**Comment by geraldzm** [comment]
Using "Ctrl+L" and then adding the % is just annoying, and then the top result is not one of the matching tabs.

A good solution is to have a fuzzy finder that let's you search by tab title or url.

This extension has the right idea: [Zen tabs](https://github.com/AntonDobrovinskiy/Zen-Tab-Search)
There should be a good native option.



https://github.com/user-attachments/assets/c1b829d1-4c9d-4768-bdf8-ddb198ac8510

**Comment by vict0rsch** [comment]
Even if a dedicated shortcut is implemented (yes please ✨), I think it would still make sense to boost open tabs up in the search results, such a situation (⬇︎) is surprising to me: the whole point of suggesting open tabs is to not open a new one so existing ones should be at the top IMHO.

E.g. with `dell` after `cmd+T` (same behaviour `cmd+L`)

<img width="781" height="426" alt="SCR-20260115-rofm" src="https://github.com/user-attachments/assets/c657e9eb-96a4-44c7-8c6d-29ff3370c287" />

but

<img width="812" height="311" alt="SCR-20260115-rohx" src="https://github.com/user-attachments/assets/d6aa53c9-f70f-4b62-b847-e9e70d67b3a0" />

In any case, a huge thanks to the Zen devs and contributors 💪

**Comment by lukebarton** [comment]
I'm using [Vimium](https://addons.mozilla.org/en-GB/firefox/addon/vimium-ff/) which lets me use shift+T (when not focused on a text input) to search open tabs.

<img width="1666" height="1294" alt="CleanShot 2026-02-01 at 23 55 04@2x" src="https://github.com/user-attachments/assets/8ffcf53d-2845-4301-8282-8dcd9f43e08b" />
